### PR TITLE
Refactor(eos_designs): structured_config for core_interfaces_and_l3_edge router_bgp

### PR DESCRIPTION
**Pull request by Vibhu-gslab**
## Change Summary

Refactor eos_designs structured_config for core_interfaces_and_l3_edge router_bgp

## Related Issue(s)

Fixes #

## Component(s) name

`arista.avd.eos_designs`

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py [hunk 1]
```diff
@@ -36,40 +35,32 @@ def router_bgp(self: AvdStructuredConfigCoreInterfacesAndL3EdgeProtocol) -> dict
...
             # RFC5549
             if self.inputs.underlay_rfc5549 and p2p_link.routing_protocol != "ebgp":
-                neighbor_interfaces.append({"name": p2p_link_data["interface"], **neighbor})
+                self.structured_config.router_bgp.neighbor_interfaces.append_new(
+                    name=p2p_link_data["interface"],
+                    remote_as=p2p_link_data["peer_bgp_as"],
+                    peer=p2p_link_data["peer"],
+                    description=p2p_link_data["peer"],
+                    peer_group=self.inputs.bgp_peer_groups.ipv4_underlay_peers.name,
+                )
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py:53]:
```suggestion
            if self.inputs.underlay_rfc5549 and p2p_link.routing_protocol != "ebgp":
                neighbor_interface = EosCliConfigGen.RouterBgp.NeighborInterfacesItem()
                neighbor_interface._update(name=p2p_link_data["interface"], **neighbor._as_dict())
                self.structured_config.router_bgp.neighbor_interfaces.append(neighbor_interface)
```

File: python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py [hunk 2]
```diff
@@ -36,40 +35,32 @@ def router_bgp(self: AvdStructuredConfigCoreInterfacesAndL3EdgeProtocol) -> dict
                 msg = f"{self.data_model}.p2p_links.[].as or {self.data_model}.p2p_links_profiles.[].as"
                 raise AristaAvdInvalidInputsError(msg)
 
-            neighbor = {
-                "remote_as": p2p_link_data["peer_bgp_as"],
-                "peer": p2p_link_data["peer"],
-                "description": p2p_link_data["peer"],
-                "peer_group": self.inputs.bgp_peer_groups.ipv4_underlay_peers.name,
-            }
+            neighbors = EosCliConfigGen.RouterBgp.NeighborsItem(
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py]:
```suggestion
            neighbor = EosCliConfigGen.RouterBgp.NeighborsItem(
```

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py]: Need to change at other places too.

Comment (Vibhu-gslab) [reply python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py]: Fixed!

File: python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py (hunk 1 above)
Comment (Vibhu-gslab) [reply python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py:53]: Fixed!

**Review by laxmikantchintakindi — APPROVED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml [hunk 3]
```diff
@@ -16,14 +16,19 @@ l3_edge:
...
       id: 2
       profile: profile2
+      as: [65000, 65002]
+      ip: [192.168.0.2/31, 192.168.0.3/31]
 
 expected_error_message: >-
-  Found duplicate objects with conflicting data while generating configuration for Ethernet Interfaces defined under l3_edge p2p_link.
-  {'name': 'ethernet1', 'peer': 'peer2', 'peer_interface': 'ethernet2', 'description': 'P2P_peer2_ethernet2'} conflicts with
-  {'name': 'ethernet1', 'peer': 'peer1', 'peer_interface': 'ethernet1', 'description': 'P2P_peer1_ethernet1'}.
+  Found duplicate objects with conflicting data while generating configuration for Neighbors. {'ip_address': '192.168.0.3',
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml]: This is detecting duplicate BGP neighbors now, where before it was checking for duplicate interfaces. Please revert and add another test for duplicate neighbors. Thanks.

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml [hunk 4]
```diff
@@ -33,6 +33,7 @@ core_interfaces:
       interfaces: [ Ethernet1, Ethernet2 ]
       profile: isis_bb_profile
       ip: ["unnumbered loopback0", "unnumbered loopback0"]
+      as: [65000, 65002]
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml:36]: why do we need to add AS and IP info here?

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml]: same for core-interface variant

File: python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py [hunk 5]
```diff
@@ -36,40 +35,26 @@ def router_bgp(self: AvdStructuredConfigCoreInterfacesAndL3EdgeProtocol) -> dict
...
+                remote_as=p2p_link_data["peer_bgp_as"],
+                peer=p2p_link_data["peer"],
+                description=p2p_link_data["peer"],
+                peer_group=self.inputs.bgp_peer_groups.ipv4_underlay_peers.name,
+            )
 
             # RFC5549
             if self.inputs.underlay_rfc5549 and p2p_link.routing_protocol != "ebgp":
-                neighbor_interfaces.append({"name": p2p_link_data["interface"], **neighbor})
+                self.structured_config.router_bgp.neighbor_interfaces.append_new(name=p2p_link_data["interface"], **neighbor._as_dict())
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py]: We should avoid going to dict at all.
It will be simpler to just build the full neighbor interface here setting the 5 arguments directly.
And then remove the `neighbor = ` stuff below this `if` so it only runs when building a regular neighbor.

Alternative is to use _cast_as

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml (hunk 4 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml:36]: We were getting the below errors and molecule was failing

```
{"changed": false, "msg": "core_interfaces.p2p_links.[].as or core_interfaces.p2p_links_profiles.[].as”}
{"changed": false, "msg": "core_interfaces.p2p_links.[].ip, .subnet or .ip_pool"}
```

File: python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py (hunk 5 above)
Comment (Vibhu-gslab) [reply python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/router_bgp.py]: Fixed!

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml (hunk 3 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interface-l3-edge.yml]: Fixed error message.

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml (hunk 4 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/duplicate-interfaces-core-interfaces-with-port-channel.yml:36]: it should be possible to not set AS so lets make sure the behavior is not changing

**Comment by Vibhu-gslab** [comment]
Moved to PR https://github.com/aristanetworks/avd/pull/5087 as discussed.
